src/metabase/util/i18n/impl.clj

```diff
@@ -203,31 +203,25 @@
              (log/errorf e "Invalid format string %s" (pr-str format-string))
              format-string)))))))
 
-;; We can't fetch the system locale until the application DB has been initiailized. Once that's done, we don't need to
-;; do the check anymore -- swapping out the getter fn with the simpler one speeds things up substantially
-(def ^:private site-locale-from-setting-fn
-  (atom
-   (fn []
-     (when-let [db-is-set-up? (resolve 'metabase.db/db-is-set-up?)]
-       (when (and (bound? db-is-set-up?)
-                  (db-is-set-up?))
-         (when-let [get-value-of-type (resolve 'metabase.models.setting/get-value-of-type)]
-           (when (bound? get-value-of-type)
-             (let [f (fn [] (get-value-of-type :string :site-locale))]
-               (reset! site-locale-from-setting-fn f)
-               (f)))))))))
+(def ^:private ^:dynamic *in-site-locale-from-setting*
+  "Whether we're currently inside a call to [[site-locale-from-setting]], so we can prevent infinite recursion."
+  false)
 
 (defn site-locale-from-setting
-  "Fetch the value of the `site-locale` Setting.
-  When metabase is shutting down, we need to log some messages after the db connection is closed, so we keep around a
-  cached-site-locale for that purpose."
+  "Fetch the value of the `site-locale` Setting, or `nil` if it is unset."
   []
-  (let [cached-site-locale (atom "en")]
```

Comment:
This `cached-site-locale` introduced #21831 actually doesn't cache anything if you look at the code closely, because we're creating a new atom every time you call the function. So effectively it's just a try-catch that always returns `"en"` in the catch. We have other code that decides `"en"` is the fallback/default, so it's more appropriate to define that only once and return `nil` here on error. This would also return `nil` if the Setting is unset.